pkg/gadgets/internal/tchandler/tracer.go

```diff
@@ -0,0 +1,264 @@
+	containerutils "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/netnsenter"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang -cflags ${CFLAGS} dispatcher ./bpf/dispatcher.bpf.c -- -I./bpf/ -I../socketenricher/bpf
+
+type attachment struct {
+	dispatcher dispatcherObjects
+	filter     *tc.Object
+
+	// users keeps track of the users' pid that have called Attach(). This can
+	// happen for two reasons:
+	// 1. several containers in a pod (sharing the netns, and hence the networking interface)
+	// 2. pods with networkHost=true
+	// In both cases, we want to attach the BPF program only once.
+	users map[uint32]struct{}
+}
+
+func (t *Handler) closeAttachment(a *attachment) {
+	if a.filter != nil {
+		t.tcnl.Filter().Delete(a.filter)
+	}
+	a.dispatcher.Close()
+}
+
+type Handler struct {
+	dispatcherMap *ebpf.Map
+	// key: network interface name on the host side
+	// value: attachment
+	attachments map[string]*attachment
+
+	// socket to talk to netlink
+	tcnl *tc.Tc
+
+	sense AttachmentSense
+}
+
+func NewHandler(sense AttachmentSense) (*Handler, error) {
+	var err error
+	var tcnl *tc.Tc
+
+	// We need to create the client on the host network namespace, otherwise it's not able to
+	// create the qdisc and filters.
+	err = netnsenter.NetnsEnter(1, func() error {
+		// Setup tc socket for communication with the kernel
+		tcnl, err = tc.Open(&tc.Config{})
+		if err != nil {
+			return fmt.Errorf("opening rtnetlink socket: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	t := &Handler{
+		attachments: make(map[string]*attachment),
+		tcnl:        tcnl,
+		sense:       sense,
+	}
+
+	// Keep in sync with tail_call map in bpf/dispatcher.bpf.c
+	dispatcherMapSpec := ebpf.MapSpec{
+		Name:       "tail_call",
+		Type:       ebpf.ProgramArray,
+		KeySize:    4,
+		ValueSize:  4,
+		MaxEntries: 1,
+	}
+	t.dispatcherMap, err = ebpf.NewMap(&dispatcherMapSpec)
+	if err != nil {
+		return nil, fmt.Errorf("creating tail_call map: %w", err)
+	}
+	return t, nil
+}
+
+func (t *Handler) newAttachment(pid uint32, iface *net.Interface, netns uint64, sense AttachmentSense) (_ *attachment, err error) {
+	a := &attachment{
+		users: map[uint32]struct{}{pid: {}},
+	}
+	defer func() {
+		if err != nil {
+			t.closeAttachment(a)
+		}
+	}()
+
+	dispatcherSpec, err := loadDispatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	consts := map[string]interface{}{
+		"current_netns": uint32(netns),
+	}
+	if err := dispatcherSpec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("RewriteConstants while attaching to pid %d: %w", pid, err)
+	}
+
+	// We create the clsact qdisc and leak it. We can't remove it because we'll break any other
+	// application (including other ig instances) that are using it.
+	if _, err = createClsActQdisc(t.tcnl, iface); err != nil {
```

Comment:
I haven't found a solution for it. IMO it's better to leak the qdisc than breaking other applications using it.

Reply:
AFAIU it's only possible to have a single clsact filter per interface:

```bash
$ sudo tc qdisc add dev $IFACE handle 0x2 clsact
Error: Exclusivity flag on, cannot modify.

$ sudo tc qdisc add dev $IFACE parent 1 clsact
Error: "clsact" is a duplicate parent ID
```

and according to the commit implementing it on the kernel (https://lwn.net/Articles/671458/):

> Since both piggyback on TC_H_CLSACT, only one of them (ingress/clsact) can exist per netdevice, and thus ingress qdisc specific behaviour can be retained for user space